### PR TITLE
Feat: support for dynamic placeholders

### PR DESCRIPTION
**Pull request by CatHood0**
## Description

Added `placeholderConfig` parameter to allow us to build placeholders dynamically depending on what attribute we want and `cursorPlaceholderConfig` that allows to the devs add a placeholder at the right (if we are using a LTR language) of the cursor (only when the paragraph and it does not contains any block attribute or selection)

For example, if we only want to display placeholders for `Headers` and `Lists`, then we would set it up like this:
```dart
configurations: QuillEditorConfigurations(
  // We can use this instance to avoid use this functionality:
  // CursorPlaceholderConfig.noPlaceholder()
  // Or, if we want to add our custom settings, we can configure manually
  // CursorPlaceholderConfig(text: "our_text'', textStyle: TextStyle(), show: true, offset: Offset(0, 2));
  // At this example case, we use this instance (just for showcases)
  cursorPlaceholderConfig: CursorPlaceholderConfig.basic(), 
  placeholderConfig:
    PlaceholderConfig(
       builders: {
         Attribute.header.key: (Attribute attr, style) {
           final values = [30, 27, 22];
           final level = attr.value as int?;
           if (level == null) return null;
           final fontSize = values[ (level - 1 < 0 || level - 1 > 3 ? 0 : level - 1)];
           return PlaceholderArguments(
             placeholderText: 'Header $level',
             style: TextStyle(fontSize: fontSize.toDouble())
                 .merge(style.copyWith(color: Colors.grey)),
             );
          },
          Attribute.list.key: (Attribute attr, style) {
            return PlaceholderArguments(
              placeholderText: 'List item',
              style: style.copyWith(color: Colors.grey),
            );
          },
      },
   ),
),
```

#### Web

https://github.com/user-attachments/assets/23e79d41-b194-41e4-aba4-839be1ec7724

#### Desktop

https://github.com/user-attachments/assets/1a36f67d-6ab7-4669-bf1c-4269476fcc7c

#### Android

https://github.com/user-attachments/assets/13d33075-d5ab-401b-bbb1-68af36900ea6

## Related Issues

_N/A_

- [x] ✨ **New feature:** Adds new functionality without breaking existing features.
- [ ] 🛠️ **Bug fix:** Resolves an issue without altering current behavior.
- [ ] 🧹 **Code refactor:** Code restructuring that does not affect behavior.
- [ ] ❌ **Breaking change:** Alters existing functionality and requires updates.
- [ ] 🧪 **Tests:** Adds new tests or modifies existing tests.
- [ ] 📝 **Documentation:** Updates or additions to documentation.
- [ ] 🗑️ **Chore:** Routine tasks, or maintenance.
- [ ] ✅ **Build configuration change:** Changes to build or deploy processes.

# TODO

- [x] We must also make a cursor placeholder that appears at the same offset of the caret on any empty line (if the line has not block or embed attributes)
- [x] We must make possible add custom block attributes keys (we need to add some validations because we don't want to add any non block attribute)
- [x]  Make this feature completely optional.
- [x]  We need to add support for web
- [x] Rename classes related to placeholders (only those added by this PR).
- [x] Fix placeholder position on mobile devices.
- [x] Provide a way to calculate correctly the offset of the placeholder (only for cursor)
- [x] Test it on all platforms (this is an issue by itself because i don't have IOS or Mac devices) 
- [x] Test if the feature could cause any issue if we have a line with several block attributes applied
- [x] Test cursor placeholder
- [x]  Add documentation about this feature.

### CODE REVIEW

**Review by singerdmx — APPROVED** [review]

File: lib/src/editor/widgets/text/text_line.dart [hunk 1]
```diff
@@ -266,6 +269,27 @@ class _TextLineState extends State<TextLine> {
   ) {
     if (nodes.isEmpty && kIsWeb) {
       nodes = LinkedList<Node>()..add(leaf.QuillText('\u{200B}'));
+    } else if (nodes.isEmpty) {
```
Comment (CatHood0) [inline lib/src/editor/widgets/text/text_line.dart:272]: I notice that i don't add support for web

**Comment by EchoEllet** [comment]
I have noticed we're adding new features and then introduce breaking changes after short time to fix a related issue so we probably want to mark the new features as experimental for a while then make them stable similar to how it's done on other project. 

For example, the `SpellCheckerService` needs to be completely rewritten. It should abstract the source of checking the spell of words and not a service for managing a client-side spell checker. The user may want to use server-side solution or a services that provide web API so they can send GET request and get the response back, or they might want to use method channel to access the system spell checker service, the other functions such as `dispose` and the others shouldn't be part of the interface instead should be only managed by the client side spell checker service provided by the library itself, or maybe we can provide helper class for that, we should only have one method that is a future similarly to Flutter, and that's a breaking change, there are ways make it backward compatible but it's not as clean as removing them and can be confusing. 

We can use the `experimental` annotation from the`meta` package. It doesn't work similarly to how it works on other languages so the user can compile without opting in.

This PR doesn't seem to have similar issues, but just in case so, we don't have to deprecate them and support them for a while.

**Review by AtlasAutocode — COMMENTED** [review]
> I have noticed we're adding new features and then introduce breaking changes after short time to fix a related issue

This morning, I received 10 emails with incremental changes. Most of the changes seem to be addressing minor changes to code that could have been implemented before submitting. I work in a branch of my personal fork, test, implement minor changes and only submit when all is done, code is complete and stable.

I agree with @EchoEllet that new features should be complete before submitting. Bugs and errors in edge cases are inevitable but code needs to be tested before submitting. Code used by our userbase must be **stable and complete**.

I am sorry to be critical because you are working so hard at improving the project. But there are an increasing number of critical issues that are not being addressed. As a couple of examples:

> For example, the SpellCheckerService needs to be completely rewritten.

Spell checking might be very important for some users, but the penalty is too high. The implementation is completely useless for me as I need a highly technical vocabulary that is never going to be implemented.

> Embedding Table

Still not implemented? Please fix or remove. Or at least mark as experimental so users know to avoid. This is a perfect example of working in a private branch until the feature is complete and ready for general use.

I am having a bad day. I apologize if my comments are too harsh. I am getting frustrated.

> placeholder appear on any empty line

NO!!!

P.S. I have not reviewed the code changes.

**Comment by EchoEllet** [comment]
Thank @AtlasAutocode for the feedback, I do agree that we should test our changes in `dev` branch and push them as pre-release since the testing is very manual. Marking them as experimental is the least we can do.

We should be more cautious when introducing new changes since medium-big projects are depending on this project. User expect the changes to be stable in the `master` or the stable release.

This is a mistake I made in 2023 when working on this repository since I didn't even have the time for writing a good commit messages or explaining my changes. Now I'm introducing much less features and working less often than before but all of my changes are much more stable now. At the end of 2023, I only had 1 hour in a month.

For now, I suggest making most of our new changes in completely separate files to make it easier to keep track of things, we don't want to make the `QuillRawEditorState` and `QuillController` and related files 40% bigger.

**Comment by CatHood0** [comment]
> We can use the experimental annotation from themeta package. It doesn't work similarly to how it works on other languages so the user can compile without opting in.

Thanks for the suggestion. I'll take it into consideration.

> This PR doesn't seem to have similar issues, but just in case so, we don't have to deprecate them and support them for a while like the [enableMarkdownStyleConversion](https://github.com/singerdmx/flutter-quill/blob/master/lib%2Fsrc%2Feditor%2Fconfig%2Feditor_configurations.dart#L63-L65).

You are correct, there are no issues related to placeholders. However, I thought we could add this functionality to improve how we view elements that have attributes in blocks, such as lists, titles, and even blockquotes that when empty, do not display any text, which is sometimes confusing (at least in my case it has been).

I take [BlockNote](https://www.blocknotejs.org/) editor as an example of how this feature should be maded:

Captura de pantalla_2024-09-18_11-53-37

> For example, the `SpellCheckerService` needs to be completely rewritten. It should abstract the source of checking the spell of words and not a service for managing a client-side spell checker. The user may want to use server-side solution or a services that provide web API so they can send GET request and get the response back, or they might want to use method channel to access the system spell checker service, the other functions such as dispose and the others shouldn't be part of the interface instead should be only managed by the client side spell checker service provided by the library itself, or maybe we can provide helper class for that, we should only have one method that is a future similarly to Flutter, and that's a breaking change, there are ways make it backward compatible but it's not as clean as removing them and can be confusing.

I will take this into account to improve the problems we are having with the current spell checker. Anyway, thanks for the suggestion.

>  For example, the SpellCheckerService needs to be completely rewritten.
> 
> Spell checking might be very important for some users, but the penalty is too high. The implementation is completely useless for me as I need a highly technical vocabulary that is never going to be implemented.

You have a point. But, some devs could need this feature. I'll add it to my TODO list. Thanks for your feedback.

>     Embedding Table
> 
> Still not implemented? Please fix or remove. Or at least mark as experimental so users know to avoid. This is a perfect example of working in a private branch until the feature is complete and ready for general use.

I think the same. Table should not be available for users. I'll make a PR to disable it by now because is quite experimental and it does not work as we expect. I'll working on it to improve our implementation.

>  placeholder appear on any empty line
> 
> NO!!!

It seems I expressed myself poorly. I blame my PC for not being able to properly record a sample video for this feature (no idea).

At this point, what I mean is that if we create our own placeholders using `placeholderComponentsConfiguration` whenever these lines are empty, their intended placeholder will appear. However, this is ONLY limited to the exclusive attributes i.e. title, list, code block, and blockquote. Any other paragraph-like or inline attributes (alignment, line-height, direction and similar attributes are also ignored to avoid putting unwanted placeholders)

Now, we have the case of the placeholder that appears at the cursor level. This is limited to only appearing IF the line is empty and it does not have any block or embedded attributes. That is, it is a full-fledged paragraph.

Needless to say, by default, none of these features are enabled. It is up to the user to configure them. If they are not desired, one can simply pass null, or not call them in the constructor.

> [!NOTE]
> If you have time, please, check [BlockNote](https://www.blocknotejs.org/). This feature works more similarly like the palceholders of this editor.

**Comment by EchoEllet** [comment]
> I will take this into account to improve the problems we are having with the current spell checker. Anyway, thanks for the suggestion.

I do suggest to discuss it before attempting to introduce any changes, I do plan to at least try to use the `DefaultSpellCheckerService` of Flutter, if it doesn't work will implement it using `quill_native_bridge`.

**Review by singerdmx — APPROVED** [review]

**Comment by EchoEllet** [comment]
Have updated my comment above, sending this comment for notification since updates doesn't send notification, will delete it after a short time to ensure a notification has been received to at least in the email. In case you're watching this repo without subscribing to the issue, this comment might be considered as spam.

**Comment by CatHood0** [comment]
> Sure but not all issues can be encountered by testing the feature in working time. Today I was able to detect an issue on Android that only occurs when copying an image from another app and when pasting it, then closing/restarting the app and attempting to paste it once again.

You have a point, you can't always detect problems until a user uses the editor.

Something quite similar happened to me with an app (which has its own repo on github) that I was helping with a while back. Apparently, if you typed certain characters and then pressed space (sort of like SpaceShorcutEvents) the app instead of displaying what it expected (a header), it displayed an error block. Apparently it didn't perform a validation when executing the event, and that error occurred (keep in mind that I was just testing out of curiosity).

**Comment by CatHood0** [comment]
# Selection issue

I noticed that there is a pretty serious bug in certain cases where the container does not record an offset that the current position does.

It's weird, I don't even know why it happens, but I noticed that this happens when we use the attributes: indent, align or line-height.

### Steps to reproduce (I have tried this by completely disabling the feature but this error still appears):

* Insert a header at the beginning of the document (write whatever you want)
* Create a new line
* ​​Below the new line you created in the previous step, apply the header attribute first and write something.
* Apply the indent attribute to this last header and you will see how this happens

I can't say exactly if this happens in the master branch, because I tried this, but no similar error appeared.

I will have to take my time to solve this with the editor and the selection.

https://github.com/user-attachments/assets/f00d5df8-f035-4469-ac38-cfb9a665d991

**Comment by CatHood0** [comment]
@singerdmx I've noticed that if we update a node with the align or indent attribute, this node will take the offset 0 for a short period (when we put the focus back or do a hot reload, this problem disappears), this same period is used by the editor to get the node (in this case the QuillContainer that has the position we tell it) and here the problem occurs. The position says that we are, for example, at the beginning of a line being this at position 15, but the node's offset is 0 (imagine that originally this node had and has starts at offset 15), that is, it will look for the first node which results in it probably not finding our current selection since it is likely that the first node does not contain such an offset with respect to the position we have.

Can you explain to me what parts I could start checking to be able to solve this?

It is not related to the feature itself, but it affects it.

**Comment by CatHood0** [comment]
I will open a separare issue for the problem with indent and make its own PR.

File: lib/src/editor/widgets/cursor.dart [hunk 2]
```diff
@@ -325,6 +331,32 @@ class CursorPainter {
...
+        );
+        placeholderPainter!
+          ..layout()
+          ..paint(canvas, offset + Offset(3.5, (style.height ?? 2) / 2));
+      }
+    }
+  }
+
+  bool _isNodeInline(Line node) {
+    for (final attr in node.style.attributes.values) {
```
Comment (CatHood0) [inline lib/src/editor/widgets/cursor.dart:356]: I notice of this code could not be here. I will make this part of an extension.

**Comment by CatHood0** [comment]
I apologize for the inactivity I have had. Time has been quite tight and it has been difficult for me to continue contributing. 

In one or two days I will start again, resume activity and continue with the PRs that I have active

**Comment by EchoEllet** [comment]
> I apologize for the inactivity I have had. Time has been quite tight and it has been difficult for me to continue contributing.

No need to apologize. Even if you don't continue, you've contributed well, and I'm grateful.

> In one or two days I will start again, resume activity and continue with the PRs that I have active

There's no urgency, so start whenever you're ready and want to.

**Comment by EchoEllet** [comment]
Same as [previous comment](https://github.com/singerdmx/flutter-quill/pull/2274#issuecomment-2476135431). How do you prefer to handle those conflicts?

**Review by EchoEllet — COMMENTED** [review]
Thank you for working on this.

Please refer to [Flutter Documentation style](https://github.com/flutter/flutter/blob/master/docs/contributing/Style-guide-for-Flutter-repo.md#documentation-dartdocs-javadocs-etc). Most of the old / original / legacy codebase, most packages and projects follow this style.

Documentation is important, I prefer if feature X doesn't exist if it could have poor documentation, and it's one of the main reasons we don't understand most of the codebase. I haven't done it in a good way previously but the goal is to improve the new changes.

File: example/lib/main.dart [hunk 3]
```diff
@@ -147,6 +142,35 @@ class _HomePageState extends State<HomePage> {
...
+                        );
+                      },
+                      Attribute.list.key: (Attribute attr, style) {
+                        return PlaceholderTextBuilder(
+                          placeholderText: 'List item',
+                          style: style.copyWith(color: Colors.grey),
+                        );
+                      },
+                    },
+                  ),
```
[the remark this answers is not in the document]
Comment (CatHood0) [reply example/lib/main.dart:173]: This was an accidental file addition. I don't know why, but no matter how many times I did a `git restore example/lib/main.dart`, it still didn't get removed from the files added to push in a commit. I'll restore it to its original version soon.

File: lib/src/editor/config/editor_config.dart [hunk 4]
```diff
@@ -137,6 +141,39 @@ class QuillEditorConfig {
...
   final List<SpaceShortcutEvent> spaceShortcutEvents;
 
+  /// Whether the line is empty, this let us add a placeholder
+  ///
+  /// _Note: these placeholders are limited only to lines with block styles_
+  ///
+  /// ### Example
+  ///
+  /// Assume that you want to show only a placeholder text for the header items,
+  /// so, we only need to configure `placeholderComponentsConfiguration` like:
```
[the remark this answers is not in the document]
Comment (CatHood0) [reply lib/src/editor/config/editor_config.dart:151]: I haven't focused on improving the documentation yet. As you said before, I will follow the documentation style you mentioned, however, my priority at the moment is to check that all the functionality is working as it should. Once I can see that everything is working correctly, I will take the time to create more documentation (not only adding doc comments but also adding files for the official documentation of the package)

File: lib/src/editor/raw_editor/builders/placeholder/placeholder_builder_internal.dart [hunk 5]
```diff
@@ -0,0 +1,120 @@
+// This file is only for internal use
+import 'package:flutter/material.dart';
+import 'package:meta/meta.dart';
+import '../../../../document/attribute.dart' show Attribute, AttributeScope;
+import '../../../../document/nodes/line.dart';
+import 'placeholder_configuration.dart';
+
+/// This is the black list of the keys that cannot be
+/// used or permitted by the builder
```
[the remark this answers is not in the document]
Comment (CatHood0) [reply lib/src/editor/raw_editor/builders/placeholder/placeholder_builder_internal.dart:9]: Thanks for the advice.

File: lib/src/editor/raw_editor/builders/placeholder/placeholder_builder_internal.dart [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+  Attribute.align.key,
+  Attribute.direction.key,
+  Attribute.lineHeight.key,
+  Attribute.indent.key,
+  ...Attribute.inlineKeys,
+  ...Attribute.ignoreKeys,
+]);
+
+@experimental
+@internal
```
[the remark this answers is not in the document]
Comment (CatHood0) [reply lib/src/editor/raw_editor/builders/placeholder/placeholder_builder_internal.dart:21]: The addition of `_internal.dart` was about a month ago while we were discussing how we should format files that were only for exclusive use within the package. I'll fix it soon.

File: lib/src/editor/raw_editor/builders/placeholder/placeholder_builder_internal.dart [hunk 7]
```diff
@@ -0,0 +1,120 @@
+// This file is only for internal use
+import 'package:flutter/material.dart';
+import 'package:meta/meta.dart';
+import '../../../../document/attribute.dart' show Attribute, AttributeScope;
+import '../../../../document/nodes/line.dart';
+import 'placeholder_configuration.dart';
+
+/// This is the black list of the keys that cannot be
+/// used or permitted by the builder
+// ignore: unnecessary_late
```
[the remark this answers is not in the document]
Comment (CatHood0) [reply lib/src/editor/raw_editor/builders/placeholder/placeholder_builder_internal.dart:10]: I probably added it while making other changes and forgot about it. It's fixed now.

**Comment by EchoEllet** [comment]
Please consider creating issues before working on them because I'm not with the idea of adding this feature or any new similar features. The editor is just not stable.

It was much better before, and so far, we haven't addressed many regressions affecting production apps. Users have already started to migrate away from the project and use alternatives.

See #2445 issue as one of mant examples, they are also many major bugs that aren't being reported.

Adding new features that cause issues will require others to take responsibility for maintaining them. We don't have enough tests, and we have already discussed this many times.

I prefer stability, especially with the current state, even if the progress is slower. If we can't fix the current issues, then it's better not to make the project any less stable.

I will create one issue for such discussions because I don't want to repeat it everywhere. 

@singerdmx
I will leave the decision to you as I don't have a very strong opinion about the close.

Adding features that we're not certain about, then fixing them later once users start reporting them comes at a high price.

**Comment by CatHood0** [comment]
@EchoEllet I didn't plan on adding this feature at the moment. I just wanted to verify that everything was in order and finish the changes. I have the same opinion on this point, I would prefer this feature to be merged after solving most of the bugs we have at the moment.

@singerdmx by now, i would prefer that this PR remain unmerged for now, and wait for a much more stable version without so many issues of the project.

**Comment by EchoEllet** [comment]
>  I would prefer this feature to be merged after solving most of the bugs we have at the moment.

There is no ETA for a more stable version. Last year, we discussed this, and it's seems everyone existed and has high expectations, but all that happens is that the magnifier was introduced and fixed, which simply makes the experience similar or even less stable than before.

> I would prefer this feature to be merged after solving most of the bugs we have at the moment.

Some users are using older versions and sticking to them due to such issues, and this is **not the kind of developer and user experience I'm looking forward to**.

And because of that, we're not certain if this PR will ever get merged, which is why I suggest creating issues for future PRs. I'm mainly concerned we end up closing many PRs that would discourage contributors with the time they have already spent.

**Comment by CatHood0** [comment]
I will close the PR as not only are we not in a position to add new features at the moment, but it will also likely be a long time before it is even merged into the stable branch.
